packages/grid/_modules_/grid/hooks/features/rowGrouping/useGridRowGrouping.tsx

```diff
@@ -0,0 +1,481 @@
+      if (currentModel.includes(field)) {
+        return;
+      }
+
+      const cleanGroupingIndex = groupingIndex ?? currentModel.length;
+
+      const updatedModel = [
+        ...currentModel.slice(0, cleanGroupingIndex),
+        field,
+        ...currentModel.slice(cleanGroupingIndex),
+      ];
+
+      apiRef.current.setRowGroupingModel(updatedModel);
+    },
+    [apiRef],
+  );
+
+  const removeRowGroupingCriteria = React.useCallback<
+    GridRowGroupingApi['removeRowGroupingCriteria']
+  >(
+    (field) => {
+      const currentModel = gridRowGroupingModelSelector(apiRef.current.state);
+      if (!currentModel.includes(field)) {
+        return;
+      }
+      apiRef.current.setRowGroupingModel(currentModel.filter((el) => el !== field));
+    },
+    [apiRef],
+  );
+
+  const setRowGroupingCriteriaIndex = React.useCallback<
+    GridRowGroupingApi['setRowGroupingCriteriaIndex']
+  >(
+    (field, targetIndex) => {
+      const currentModel = gridRowGroupingModelSelector(apiRef.current.state);
+      const currentTargetIndex = currentModel.indexOf(field);
+
+      if (currentTargetIndex === -1) {
+        return;
+      }
+
+      const updatedModel = [...currentModel];
+      updatedModel.splice(targetIndex, 0, updatedModel.splice(currentTargetIndex, 1)[0]);
+
+      apiRef.current.setRowGroupingModel(updatedModel);
+    },
+    [apiRef],
+  );
+
+  useGridApiMethod<GridRowGroupingApi>(
+    apiRef,
+    {
+      setRowGroupingModel,
+      addRowGroupingCriteria,
+      removeRowGroupingCriteria,
+      setRowGroupingCriteriaIndex,
+    },
+    'GridRowGroupingApi',
+  );
+
+  /**
+   * EVENTS
+   */
+  const handleCellKeyDown = React.useCallback<GridEventListener<GridEvents.cellKeyDown>>(
+    (params, event) => {
+      const cellParams = apiRef.current.getCellParams(params.id, params.field);
+      if (isGroupingColumn(cellParams.field) && event.key === ' ' && !event.shiftKey) {
+        event.stopPropagation();
+        event.preventDefault();
+
+        const filteredDescendantCount =
+          gridFilteredDescendantCountLookupSelector(apiRef.current.state)[params.id] ?? 0;
+
+        const isOnGroupingCell =
+          props.rowGroupingColumnMode === 'single' ||
+          getGroupingColDefFieldFromGroupingCriteriaField(params.rowNode.groupingField) ===
+            params.field;
+        if (!isOnGroupingCell || filteredDescendantCount === 0) {
+          return;
+        }
+
+        apiRef.current.setRowChildrenExpansion(params.id, !params.rowNode.childrenExpanded);
+      }
+    },
+    [apiRef, props.rowGroupingColumnMode],
+  );
+
+  const checkGroupingColumnsModelDiff = React.useCallback<
+    GridEventListener<GridEvents.columnsChange>
+  >(() => {
+    const rowGroupingModel = gridRowGroupingSanitizedModelSelector(apiRef.current.state);
+    const lastGroupingColumnsModelApplied = sanitizedModelOnLastRowPreProcessing.current;
+
+    if (!isDeepEqual(lastGroupingColumnsModelApplied, rowGroupingModel)) {
```

Comment:
The `isDeepEqual` shouldn't be necessary if reselect allowed to check only if part of the state has changed. Currently, it checks if `state === prevState` which is always false. I tried https://github.com/reduxjs/reselect#customize-equalitycheck-for-defaultmemoize to avoid an infinite loop in the variable row height but didn't get nothing concrete.

Reply:
I think it was usefull when the model was not stored in state but derived from the columns.
Because then, modifying a column not related to grouping was causing a new reference of the selector and therefore a new useless application of the grouping process.
But with the model now stored in state it should not be useful.

Reply:
I don't have a solution yet. What I noticed is that any selector, that has a result function, the reference to its value will change on each rerender, because the reference to the whole state (`apiRef.current.state`) has changed.